WIP: add replace-provider capability

mdb opened the pull request:
This is a WIP attempt to address issue #144 by providing a `replace-provider` action.

mdb commented on tfmigrate/state_replace_provider_action_test.go:
```diff
@@ -0,0 +1,38 @@
+
+	expected := "Could not load the schema for provider\nregistry.tfmigrate.io/hashicorp/null"
+	if !strings.Contains(err.Error(), expected) {
+		t.Fatalf("expected migrator plan error to include: %s", expected)
```
☝️ There may be a better way to author this test. In this case, because we've replaced the legitimate `registry.terraform.io/hashicorp/local` provider with an illegitimate `registry.tfmigrate.io/hashicorp/local` provider `plan` is expected to error, but we can verify that it does so with the expected illegitimate provider replacement.

mdb commented on tfmigrate/state_replace_provider_action_test.go:
```diff
@@ -0,0 +1,54 @@
+
+	if !strings.Contains(planErr.Error(), expected) {
+		t.Fatalf("expected migrator plan error to include: %s; received: %s", expected, err)
+	}
```
☝️ There may be a better way to author this test. In this case, because we've replaced the legitimate `registry.terraform.io/hashicorp/local` provider with an illegitimate `registry.tfmigrate.io/hashicorp/local` provider, `plan` is expected to error, but we can verify that it does so with the expected illegitimate provider replacement, in effect validating that `replace-provider` did its job.

minamijoyo replied:
Hmmm, this test is quite tricky. I'll investigate if there is any way to test for a happy path.

minamijoyo replied:
I couldn't find the best solution. Let me share possible options:

(1) Use a registry proxy such as https://github.com/jasonwbarnett/terraform-registry-proxy
Pros: A general solution for all providers
Cons: Require additional setup for testing

(2) Use any untrusted fork of hashicorp/null and rename to it
Pros: Easy to use
Cons: Potential supply chain attack risk

(3) Fork hashicorp/null by myself
Pros: Trust to use
Cons: Publishing a forked provider to the Terraform Registry takes a lot of work. I don't have experience in developing providers.

(4) Use a real-world example of renaming a namespace for verified providers such as hashicorp/github to integrations/github
Pros: Trust to use
Cons: Need a valid credential
The data source did not result in a plan diff even with namespace renaming, so a resource is needed. However, we do not want to create a real resource for testing. I tried to import a public repository, but it requires a valid GitHub token even a public one.

(5) Embed a legacy v0.12.x tfstate example as a test fixture and force push to remote in advance
Pros: No additional dependency
Cons: As I mentioned in #144, the test may break in the future because the behavior is undefined when an old state file format is loaded with a new version of terraform state commands.

minamijoyo replied:
As a reasonable compromise, I am leaning towards option (5) and am trying it out, but when I load a legacy tfstate created in v0.12.31 with Terraform v1.5.5, the terraform init fails.
I'm wondering if I need to run state replace-provider before terraform init. Does it work as intended in your upgrade scenario?

```
a05728d678f2:/work/tmp/pr-145/dir5# cat main.tf
resource "null_resource" "foo" {}
a05728d678f2:/work/tmp/pr-145/dir5# terraform -v
Terraform v1.5.5
on linux_arm64

Your version of Terraform is out of date! The latest version
is 1.5.6. You can update by downloading from https://www.terraform.io/downloads.html

a05728d678f2:/work/tmp/pr-145/dir5# terraform init && terraform apply -auto-approve

a05728d678f2:/work/tmp/pr-145/dir5# terraform state list
null_resource.foo

a05728d678f2:/work/tmp/pr-145/dir5# terraform providers

Providers required by configuration:
.
└── provider[registry.terraform.io/hashicorp/null]

Providers required by state:

    provider[registry.terraform.io/hashicorp/null]
```

```
a05728d678f2:/work/tmp/pr-145/dir5# cat legacy.tfstate
{
  "version": 4,
  "terraform_version": "0.12.31",
  "serial": 1,
  "lineage": "3cb3461c-0eab-a7b2-e2e0-57a93c3cc2a1",
  "outputs": {},
  "resources": [
    {
      "mode": "managed",
      "type": "null_resource",
      "name": "foo",
      "provider": "provider.null",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "4107759474886503934",
            "triggers": null
          }
        }
      ]
    }
  ]
}

a05728d678f2:/work/tmp/pr-145/dir5# terraform state push -force legacy.tfstate

a05728d678f2:/work/tmp/pr-145/dir5# terraform providers

Providers required by configuration:
.
└── provider[registry.terraform.io/hashicorp/null]

Providers required by state:

    provider[registry.terraform.io/-/null]

a05728d678f2:/work/tmp/pr-145/dir5# terraform init

Initializing the backend...
╷
│ Error: Invalid legacy provider address
│
│ This configuration or its associated state refers to the unqualified provider "null".
│
│ You must complete the Terraform 0.13 upgrade process before upgrading to later versions.
╵
```

```
a05728d678f2:/work/tmp/pr-145/dir5# terraform state replace-provider registry.terraform.io/-/null registry.terraform.io/hashicorp/null
Terraform will perform the following actions:

  ~ Updating provider:
    - registry.terraform.io/-/null
    + registry.terraform.io/hashicorp/null

Changing 1 resources:

  null_resource.foo

Do you want to make these changes?
Only 'yes' will be accepted to continue.

Enter a value: yes

Successfully replaced provider for 1 resources.

a05728d678f2:/work/tmp/pr-145/dir5# terraform providers

Providers required by configuration:
.
└── provider[registry.terraform.io/hashicorp/null]

Providers required by state:

    provider[registry.terraform.io/hashicorp/null]

a05728d678f2:/work/tmp/pr-145/dir5# terraform init

Initializing the backend...

Initializing provider plugins...
- Reusing previous version of hashicorp/null from the dependency lock file
- Using previously-installed hashicorp/null v3.2.1

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.

a05728d678f2:/work/tmp/pr-145/dir5# terraform plan
null_resource.foo: Refreshing state... [id=4107759474886503934]

No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and found no differences,
so no changes are needed.
```

mdb replied:
> I'm wondering if I need to run state replace-provider before terraform init. Does it work as intended in your upgrade scenario?

I don't believe you can run `terraform state replace-provider` before first running `terraform init`. Otherwise, you'll run into `Error loading the state` error:

```
terraform state replace-provider registry.terraform.io/-/null registry.terraform.io/hashicorp/null
Error loading the state: Backend initialization required, please run "terraform init": Reason: Initial configuration of th
e requested backend "s3"
```

In my experience, when using a Terraform version >= v0.13.0 to run `terraform state replace-provider` against a legacy v0.12.31 tfstate, it's necessary to...

1. First run `terraform init`; this will fail with the `Error: Invalid legacy provider address` error you mention above. Despite the failure, it's still necessary.
2. Run `terraform state replace-provider`

minamijoyo replied:
So, to support state replace-provider -/null hashicorp/null for legacy v0.12.x tfstate described in #144, we need a special error handling in the work directory initialization step, right?
I suggest the acceptance test confirm this scenario by embedding the legacy v0.12.x tfstate as a test fixture.

minamijoyo replied:
Shouldn't we need to fix OverrideBackendToLocal as well as setupWorkDir? The error occurs during rollback but does not throw the error. It works in unintended balance, but it is not so user-friendly.

```
6d3015a26631:/work/tmp/pr-145/dir11# cat test.hcl
migration "state" "test" {
  dir = "."
  actions = [
    "replace-provider registry.terraform.io/-/null registry.terraform.io/hashicorp/null",
  ]
}

6d3015a26631:/work/tmp/pr-145/dir11# tfmigrate plan test.hcl
2023/08/28 14:52:35 [INFO] [runner] load migration file: test.hcl
2023/08/28 14:52:35 [INFO] [migrator] start state migrator plan
2023/08/28 14:52:35 [INFO] [migrator@.] terraform version: 1.5.5
2023/08/28 14:52:35 [INFO] [migrator@.] initialize work dir
2023/08/28 14:52:35 [INFO] [migrator@.] ignoring error 'Error: Invalid legacy provider address' initilizing work dir; the error is expected when using Terraform >= 0.13 with a legacy Terraform state
2023/08/28 14:52:35 [INFO] [migrator@.] get the current remote state
2023/08/28 14:52:35 [INFO] [migrator@.] override backend to local
2023/08/28 14:52:35 [INFO] [executor@.] create an override file
2023/08/28 14:52:35 [INFO] [migrator@.] creating local workspace folder in: terraform.tfstate.d/default
2023/08/28 14:52:35 [INFO] [executor@.] switch backend to local
2023/08/28 14:52:35 [INFO] [migrator@.] compute a new state
2023/08/28 14:52:36 [INFO] [migrator@.] check diffs
2023/08/28 14:52:36 [INFO] [executor@.] remove the override file
2023/08/28 14:52:36 [INFO] [executor@.] remove the workspace state folder
2023/08/28 14:52:36 [INFO] [executor@.] switch back to remote
2023/08/28 14:52:36 [ERROR] [executor@.] failed to switch back to remote: failed to run command (exited 1): terraform init -input=false -no-color -reconfigure
stdout:

Initializing the backend...

Successfully configured the backend "s3"! Terraform will automatically
use this backend unless the backend configuration changes.

stderr:

Error: Invalid legacy provider address

This configuration or its associated state refers to the unqualified provider
"null".

You must complete the Terraform 0.13 upgrade process before upgrading to
later versions.


2023/08/28 14:52:36 [ERROR] [executor@.] please re-run terraform init -reconfigure
2023/08/28 14:52:36 [INFO] [migrator] state migrator plan success!

6d3015a26631:/work/tmp/pr-145/dir11# echo $?
0
```

minamijoyo replied:
My intuition was that the error would also occur when switching the backend to local, but I wonder why it doesn't.
Switching to local with a direct aws s3 cp of a v0.12.31 tfstate doesn't cause an error, but with terraform v1.5.5, After state push, switching to local also gives an error.
Rollback to remote will give an error in either case.
I don't know why 🤷‍♂️

minamijoyo replied:
Anyway, it is better to fix the OverrideBackendToLocal function to suppress the legacy state error in both the switch and rollback phases, isn't it?

mdb replied:
👍 Good call! I took a stab at implementing your suggestion in https://github.com/minamijoyo/tfmigrate/pull/145/commits/37e28346ecda3b006e4f69dcc666b8f666814eef. How's that look?

mdb commented:
@minamijoyo Addressing issue #145 and adding `replace-provider` functionality turned out to be a bit more quirky than I anticipated, but this first pass implementation is ready for your review. Thanks!

minamijoyo commented:
@mdb Thanks for working on this! As you said, it's more complicated than I expected, and there are some points where I need to check the Terraform CLI behavior, so please give me some time to review it. Instead of returning the review results to you in a summary, I will comment on them in order as I study the areas in my spare time.

minamijoyo commented on tfexec/terraform.go:
```diff
@@ -96,6 +101,13 @@ type TerraformCLI interface {
+	// If a state argument is given, use it for the input state.
+	// It returns the given state.
+	// Unlike other state subcommands, the terraform state replace-provider
+	// command doesn't support a -state-out option; it only supports the -state option.
```
I tried it, too, and confirmed that the -state-out flag was not defined.
It is a Terraform CLI bug because the documentation and implementation do not match.
https://developer.hashicorp.com/terraform/cli/commands/state/replace-provider
https://github.com/hashicorp/terraform/blob/v1.5.5/internal/command/state_replace_provider.go

```
a05728d678f2:/work/tmp/pr-145/dir1# terraform -v
Terraform v1.5.5
on linux_arm64
+ provider registry.terraform.io/hashicorp/null v3.2.1

a05728d678f2:/work/tmp/pr-145/dir1# terraform state replace-provider -state=tmp.tfstate -state-out=out.tfstate hashicorp/null minamijoyo/null
Error parsing command-line flags: flag provided but not defined: -state-out

Usage: terraform [global options] state replace-provider [options] FROM_PROVIDER_FQN TO_PROVIDER_FQN

  Replace provider for resources in the Terraform state.

Options:

  -auto-approve           Skip interactive approval.

  -lock=false             Don't hold a state lock during the operation. This is
                          dangerous if others might concurrently run commands
                          against the same workspace.

  -lock-timeout=0s        Duration to retry a state lock.

  -ignore-remote-version  A rare option used for the remote backend only. See
                          the remote backend documentation for more information.

  -state, state-out, and -backup are legacy options supported for the local
  backend only. For more information, see the local backend's documentation.
```

mdb replied:
> It is a Terraform CLI bug because the documentation and implementation do not match.

That is my sense too, though I haven't fully researched the backstory.

minamijoyo commented on tfexec/terraform_providers_test.go:
```diff
@@ -0,0 +1,121 @@
+
+	got, err := terraformCLI.Providers(context.Background())
+	if err != nil {
+		t.Fatalf("failed to run terraform state list: %s", err)
```
The error message should be `terraform providers`, not `terraform state list`.

minamijoyo commented on tfexec/terraform_state_replace_provider.go:
```diff
@@ -0,0 +1,71 @@
+		return nil, err
+	}
+
+	v, err := c.Version(ctx)
```
It may be ideal for getting and caching the version at the time of initialization of terraformCLI. Still, the scope of refactoring would be significant, so we will postpone it for now.

minamijoyo commented on tfexec/terraform_state_replace_provider_test.go:
```diff
@@ -0,0 +1,290 @@
+					argsRe:   regexp.MustCompile(`^terraform version$`),
+					runFunc:  versionRunFunc,
+					stdout:   "Terraform v0.12.99\n",
+					exitCode: 1,
```
When the exit code of the terraform version command is 1,  it doesn't work as expected because the function returns before checking the version constraint.

minamijoyo commented on tfmigrate/state_action.go:
```diff
@@ -45,6 +45,14 @@ func NewStateActionFromString(cmdStr string) (StateAction, error) {
 		dst := args[2]
 		action = NewStateMvAction(src, dst)
 
+	case "replace-provider":
```
Could you add the replace-provider case to the above godoc comments and unit tests?

minamijoyo commented on tfmigrate/state_replace_provider_action.go:
```diff
@@ -0,0 +1,37 @@
+	destination string
+}
+
+var _ StateAction = (*StateMvAction)(nil)
```
It should be StateReplaceProviderAction, not StateMvAction, since this is the idiomatic expression for type-checking that the struct implements the interface.

minamijoyo commented on tfmigrate/state_replace_provider_action.go:
```diff
@@ -0,0 +1,37 @@
+}
+
+// StateUpdate updates a given state and returns a new state.
+// It moves a resource from source address to destination address in the same tfstate file.
```
This line also seems to have been copied from mv.

minamijoyo commented on tfmigrate/state_replace_provider_action.go:
```diff
@@ -0,0 +1,37 @@
+	// because we never restore state from the backup generated by each state action.
+	// The state replace-provider command doesn't provide a way to disable it, so we backup to /dev/null.
+	newState, err := tf.StateReplaceProvider(ctx, state, a.source, a.destination, "-backup=/dev/null", "-auto-approve")
+	return newState, err
```
Why not return the result directly without assigning it to local variables?

mdb commented:
> @mdb Thanks for working on this! As you said, it's more complicated than I expected, and there are some points where I need to check the Terraform CLI behavior, so please give me some time to review it. Instead of returning the review results to you in a summary, I will comment on them in order as I study the areas in my spare time.

Thanks @minamijoyo - I believe I've addressed the concerns you've expressed so far. I'll stay tuned for any further feedback!

mdb commented on Makefile:
```diff
@@ -32,3 +32,18 @@ testacc: build generate-plugin-cache
+		hashicorp/terraform:0.12.31 \
+			-c \
+				"terraform init && \
+				terraform apply -auto-approve"
```
☝️ This is a convenience utility for generating the 0.12.31 TF state test fixture. Perhaps this is useful in the future? If anything, I hope it offers some transparent visibility on my technique for generating the legacy TF state file.

I'm not offended if you feel this is unwarranted, though, @minamijoyo.

minamijoyo replied:
I don't think it is required, but you may leave it as is.

mdb commented on docker-compose.yml:
```diff
@@ -18,6 +18,7 @@ services:
       # From observation, although we don’t have complete confidence in the root cause,
       # it appears that localstack sometimes misses API requests when run in parallel.
       TF_CLI_ARGS_apply: "--parallelism=1"
+      TERRAFORM_VERSION: ${TERRAFORM_VERSION:-latest}
```
☝️ I believe this is necessary so certain acceptance tests can be conditionally skipped based on the Terraform version in use, based on `TERRAFORM_VERSION`.

minamijoyo replied:
Why not use SupportsStateReplaceProvider()? It directly represents the skip condition of the test.

minamijoyo replied:
OK. The test for state replace-provider requires special setups, and we want to check the version before initialization 👌

mdb commented on tfmigrate/migrator.go:
```diff
@@ -30,11 +31,21 @@ func setupWorkDir(ctx context.Context, tf tfexec.TerraformCLI, workspace string,
-		return nil, nil, err
+		acceptable := "Error: Invalid legacy provider address"
+		if supportsStateReplaceProvider && strings.Contains(err.Error(), acceptable) {
+			log.Printf("[INFO] [migrator@%s] ignoring error '%s' initilizing work dir; the error is expected when using Terraform %s with a legacy Terraform state\n", tf.Dir(), acceptable, constraints)
```
☝️ Could/should this logic be improved? For example, perhaps the `Error: Invalid legacy provider address` should only be permitted during `state replace-provider` actions, but not other actions? Doing so would require passing the action(s) to `setupWorkDir`, though. Do you have thoughts, @minamijoyo ?

minamijoyo replied:
I think it is safer to narrow down the special error-handling conditions.
Rather than passing actions directly to the setupWorkDir function, why not pass it as a bool argument, such as ignoreLegacyState?

mdb replied:
@minamijoyo Yep; we can pass a `bool` flag if you prefer. I opted not to, as that arguably redundantly distributes the need to correctly set this `bool` value across all invocations of the function (rather than consolidated the logic to a single place, within the function). However, perhaps you're right that doing so is best in this case. Stay tuned :)

mdb replied:
@minamijoyo I believe I've implemented your suggestion in https://github.com/minamijoyo/tfmigrate/pull/145/commits/67ed6758bc45d3c881e75c5c56a3f5d497913cc5. Does that seem acceptable?

minamijoyo commented on tfmigrate/migrator.go:
```diff
@@ -60,7 +70,7 @@ func setupWorkDir(ctx context.Context, tf tfexec.TerraformCLI, workspace string,
 	// override backend to local
 	log.Printf("[INFO] [migrator@%s] override backend to local\n", tf.Dir())
-	switchBackToRemoteFunc, err := tf.OverrideBackendToLocal(ctx, "_tfmigrate_override.tf", workspace, isBackendTerraformCloud, backendConfig)
+	switchBackToRemoteFunc, err := tf.OverrideBackendToLocal(ctx, "_tfmigrate_override.tf", workspace, isBackendTerraformCloud, backendConfig, supportsStateReplaceProvider)
```
@mdb Thank you for working on this!
I think we should pass ignoreLegacyStateInitErr instead of supportsStateReplaceProvider to OverrideBackendToLocal. Because the supportStateReplaceProvider depends on the Terraform version; on the other hand, the ignoreLegacyStateInitErr depends on actions. The ignoreLegacyStateInitErr is more strict to check.

mdb replied:
@minamijoyo That sounds reasonable - I pushed a change doing so!

minamijoyo reviewed (approved):
Thank you for your great work! LGTM :+1:

mdb commented:
@minamijoyo Thanks so much for the patient reviews and great feedback!

minamijoyo commented:
@mdb Thank you for your contribution! I've shipped it in v0.3.15. 🚀
